hexgame-rs/src/game/board.rs

```diff
@@ -0,0 +1,300 @@
+use super::cell::Cell;
+use super::cell::Ownership;
+
+/*
+ A board is a set of exagonal cells stacked in a 2D matrix shape.
+ It can be indexed similarly to a matrix with (x, y) coordinates.
+ The (0,0) coordinate starts at the bottom left corner of the board.
+ The x coordinate indexes horizontally (going from left to right)
+ and the y vertically (going from bottom to top)
+ The exagonal cells make so that the board is slanted in a
+ romboidal shape. The slanting goes to the left.
+ ----------
+ '         '
+  '         '
+   '         '
+     ----------
+*/
+
+#[derive(Debug)]
+pub struct Board {
+    pub dim_x: usize,
+    pub dim_y: usize,
+    pub cells: Vec<Vec<Cell>>,
+}
+
+impl Board {
+    pub fn new() -> Self {
+        Self {
+            dim_x: 0,
+            dim_y: 0,
+            cells: vec![vec![]],
+        }
+    }
+
+    pub fn new_from_dims(dim_x: usize, dim_y: usize) -> Self {
+        let mut matrix = Vec::with_capacity(dim_x);
+        for x in 0..dim_x {
+            let mut column = Vec::with_capacity(dim_y);
+            for y in 0..dim_y {
+                column.push(Cell::new(x, y));
+            }
+            matrix.push(column);
+        }
+        Self {
+            dim_x,
+            dim_y,
+            cells: matrix,
+        }
+    }
+
+    pub fn new_from_dim(dim: usize) -> Self {
+        return Board::new_from_dims(dim, dim);
+    }
+
+    fn is_empty(&self) -> bool {
+        self.dim_x == 0 || self.dim_y == 0
+    }
+
+    fn is_valid_square(&self, x: isize, y: isize) -> bool {
+        if self.is_empty() {
+            panic!("The board is empty")
+        }
+        return !(x < 0
+            || y < 0
+            || x >= self.dim_x.try_into().unwrap()
+            || y >= self.dim_y.try_into().unwrap());
```

Comment:
This is something that I have already considered in the past.
I tried to modify the types from `usize` to `isize` now and the problem of casting shifts in other parts of the code.
It turns out that having the dimension as unsigned integers is necessary to avoid casting when using the values to assign vector dimensions.
I would prefer keeping them `usize`